lib/deepcrawl.js

```diff
@@ -0,0 +1,137 @@
+'use strict';
+const BPromise = require('bluebird'),
+  schema = require('../schemas/v2.0.0'),
+  needle = BPromise.promisifyAll(require('needle')),
+  humps = require('humps'),
+  _ = require('lodash');
+
+class DeepCrawl {
+  constructor(cfg) {
+    cfg = cfg || {};
+
+    this.apiId = cfg.apiId;
+    this.apiKey = cfg.apiKey;
+    this.getSessionToken = cfg.getSessionToken;
+    this.accountId = cfg.accountId;
+
+    this.schema = cfg.schema || schema;
+    this.version = this.schema.version;
+    this.baseUrl = this.schema.baseUrl;
+  }
+
+  createEndpoints (schema) {
+    const API = {};
+    for (const resource in schema.resources) {
+      API[resource] = {};
+    }
+    return API;
+  }
+
+  generateRequest (url, options, method) {
+    // first, get the sessionToken and generate the request options and headers
+    return BPromise.resolve(options.sessionToken || this.getSessionToken())
+      .then((sessionToken) => {
+        const requestOpts = {
+          json: true,
+          headers: {
+            'X-Auth-Token': sessionToken
+          }
+        };
+
+        if (method === 'get') {
+          return needle.getAsync(url, requestOpts)
+            .then(this.handleResponse);
+        }
+
+        // convert all camelCased options to _ for deepcrawl
+        const data = humps.decamelizeKeys(options);
+        return needle[`${method}Async`](url, data, requestOpts)
+          .then(this.handleResponse);
+      });
+  }
+
+  generateMethod (requiredFields, url, method) {
+    return (options) => {
+      options = options || {};
+      if (!options.sessionToken && !this.getSessionToken) {
+        throw new Error(`"sessionToken" or "getSessionToken" method required`);
+      }
+
+      requiredFields.forEach((field) => {
+        if (!options[field] && !this[field]) {
+          throw new Error(`${field} is required`);
+        }
+        // if the url contains {field}, replace {field} with the passed
+        // in value
+        if (url.indexOf(`{${field}}`) > -1) {
+          url = url.replace(`{${field}}`, (options[field] || this[field]));
+        }
+      });
+
+      return this.generateRequest(url, options, method);     
+    };
+  }
+
+  addMethodsToEndpoints (API, schema) {
+    for (const resource in API) {
```

Comment:
I think this method should be renamed to something more general like `parseSchema` which is really all it's doing. This initial loop can simply be `for (const resource in API.resources) {` to avoid the extraneous `createEndpoints` method -- that's the whole beauty of adding that `resources` level so you no longer have to explicitly parse them out -- they are already there ;)